### PR TITLE
PSR-631 Fixed bugs

### PR DESCRIPTION
**Pull request by westwater**

### CODE REVIEW

File: app/navigation/nonsipp/package.scala [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+import eu.timepit.refined.api.{Refined, Validate}
+import eu.timepit.refined.refineV
+
+package object nonsipp {
+
+  // Given a list of indexes, this function will check for the next open index to start a new journey
+  // Will return None if the the resulting index is out of bounds of the supplied refined type
+  def findNextOpenIndex[A](indexes: List[Int])(implicit ev: Validate[Int, A]): Option[Refined[Int, A]] = {
+    val sortedIndex = indexes.sorted
```
Comment (kberec) [inline app/navigation/nonsipp/package.scala:27]: Just as an idea.... would it simplify the logic below to sort in reverse order here?

File: app/viewmodels/models/MemberState.scala [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import utils.WithName
+
+sealed trait MemberState {
+  val name: String
+}
+
+object MemberState {
+  case object Active extends WithName("Active") with MemberState
+
+  case object Deleted extends WithName("Deleted") with MemberState
```
Comment (kberec) [inline app/viewmodels/models/MemberState.scala:29]: This is a pre-requisite for soft delete, but it hasn't been used yet. Is this understanding correct?

File: app/controllers/testonly/PrintsUserAnswersController.scala [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import scala.concurrent.ExecutionContext
+
+class PrintsUserAnswersController @Inject()(
+  identifyAndRequireData: IdentifyAndRequireData,
+  val controllerComponents: MessagesControllerComponents
+)(implicit ec: ExecutionContext)
+    extends FrontendBaseController
+    with I18nSupport {
+
+  def printUserAnswers(srn: Srn): Action[AnyContent] =
```
Comment (kberec) [inline app/controllers/testonly/PrintsUserAnswersController.scala:36]: This is really useful!

File: app/controllers/nonsipp/memberdetails/SchemeMembersListController.scala [hunk 4]
```diff
@@ -75,19 +80,66 @@ class SchemeMembersListController @Inject()(
       form(manualOrUpload, membersDetails.size >= maxSchemeMembers)
         .bindFromRequest()
         .fold(
-          formWithErrors =>
-            BadRequest(
-              view(formWithErrors, viewModel(srn, page, manualOrUpload, mode, membersDetails.map(_.fullName)))
-            ),
+          formWithErrors => {
+            filterDeletedMembers(srn).map { filteredMembers =>
```
Comment (kberec) [inline app/controllers/nonsipp/memberdetails/SchemeMembersListController.scala:84]: Looks this is breaking the display of the validation errors, i.e. if you don't select a value on the radio button.

File: app/controllers/nonsipp/memberdetails/SchemeMembersListController.scala [hunk 5]
```diff
@@ -75,19 +80,66 @@ class SchemeMembersListController @Inject()(
...
+              Redirect(
+                navigator.nextPage(SchemeMembersListPage(srn, value, manualOrUpload), mode, request.userAnswers)
+              )
             }
           }
         )
     }
+
+  // merges member details with member states and filters out any soft deleted members
+  private def filterDeletedMembers(
```
Comment (kberec) [inline app/controllers/nonsipp/memberdetails/SchemeMembersListController.scala:107]: Struggling to understand why do we need to filter out on the validation error display, i.e. why is this method called from onSubmit - formWithErrors part.

File: app/viewmodels/models/MemberState.scala (hunk 2 above)
Comment (westwater) [reply app/viewmodels/models/MemberState.scala:29]: Yup thats right - I started to add it here before splitting the tickets

File: app/controllers/nonsipp/memberdetails/SchemeMembersListController.scala (hunk 5 above)
Comment (westwater) [reply app/controllers/nonsipp/memberdetails/SchemeMembersListController.scala:107]: I forgot to add to on page load aswell - the page should not display any deleted members

File: app/controllers/nonsipp/memberdetails/SchemeMembersListController.scala [hunk 6]
```diff
@@ -75,19 +80,66 @@ class SchemeMembersListController @Inject()(
...
+    srn: Srn
+  )(implicit request: DataRequest[_]): Either[Result, List[(Max300, NameDOB, MemberState)]] = {
+    val maybeMembersDetails =
+      request.userAnswers.membersDetailsMap(srn).mapKeysToIndex[Max300.Refined].getOrRecoverJourney
+    val maybeMemberStates = request.userAnswers.memberStates(srn).mapKeysToIndex[Max300.Refined].getOrRecoverJourney
+
+    maybeMembersDetails.flatMap { membersDetails =>
+      maybeMemberStates.flatMap { memberStates =>
+        if (membersDetails.size != memberStates.size) {
```
[the remark this answers is not in the document]
Comment (westwater) [reply app/controllers/nonsipp/memberdetails/SchemeMembersListController.scala:123]: fixed

File: app/controllers/nonsipp/memberdetails/SchemeMembersListController.scala (hunk 4 above)
Comment (westwater) [reply app/controllers/nonsipp/memberdetails/SchemeMembersListController.scala:87]: fixed

File: app/navigation/nonsipp/package.scala (hunk 1 above)
Comment (westwater) [reply app/navigation/nonsipp/package.scala]: simplified it a lot now

**Review by kberec — APPROVED** [review]

**Review by kberec — APPROVED** [review]

**Pull request merged**